core/src/main/scala/org/apache/spark/shuffle/hash/HashShuffleManager.scala

```diff
@@ -50,15 +53,30 @@ private[spark] class HashShuffleManager(conf: SparkConf) extends ShuffleManager
   }
 
   /** Get a writer for a given partition. Called on executors by map tasks. */
-  override def getWriter[K, V](handle: ShuffleHandle, mapId: Int, context: TaskContext)
-      : ShuffleWriter[K, V] = {
-    new HashShuffleWriter(
-      shuffleBlockResolver, handle.asInstanceOf[BaseShuffleHandle[K, V, _]], mapId, context)
+  override def getWriter[K, V](
+      handle: ShuffleHandle,
+      mapId: Int,
+      stageAttemptId: Int,
+      context: TaskContext): ShuffleWriter[K, V] = {
+    addShuffleAttempt(handle.shuffleId, stageAttemptId)
+    new HashShuffleWriter(shuffleBlockResolver, handle.asInstanceOf[BaseShuffleHandle[K, V, _]],
```

Comment:
I added some comments, but I opted against the `getWriterHelper` version b/c of the delegation inside `UnsafeShuffleManager`.  I felt like that added enough complexity to making sure you didn't double-register shuffles and cleanup still happened only once in all cases that is out-weighed the benefits.  But its not a clear win either way ... I can finish writing up that version if you think its worth comparing them both.